kernel/src/upcall.rs

```diff
@@ -83,23 +88,23 @@ pub(crate) struct Upcall {
     pub(crate) upcall_id: UpcallId,
 
     /// The application data passed by the app when `subscribe()` was called.
-    pub(crate) appdata: usize,
+    pub(crate) appdata: AppdataType,
 
     /// A pointer to the first instruction of the function in the app that
     /// corresponds to this upcall.
     ///
     /// If this value is `None`, this is a null upcall, which cannot actually be
     /// scheduled. An `Upcall` can be null when it is first created, or after an
     /// app unsubscribes from an upcall.
-    pub(crate) fn_ptr: Option<NonNull<()>>,
+    pub(crate) fn_ptr: FnPtrType,
```

Comment:
`CapabilityPtr` is the equivalent of `*const ()`, which is nullable, and I added all methods that `Option<NonNull<()>>` had for handling the NULL case so code would otherwise remain unchanged / be semantically the same.

I agree, `Option<NonNull<()>>` is better, but then we would need to be able to express "Nullable Capability" to the compiler, which we cannot do yet. This issue is called out with the type aliases, and the indirection is a simple change once compiler support is better.